### PR TITLE
expression: rewrite builtin string function: FROM_BASE64

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2116,35 +2116,37 @@ type fromBase64FunctionClass struct {
...
+	str, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
 	}
 	str = strings.Replace(str, "\t", "", -1)
 	str = strings.Replace(str, " ", "", -1)
 	result, err := base64.StdEncoding.DecodeString(str)
 	if err != nil {
-		return d, errors.Trace(err)
+		return "", true, nil
```
Comment (zz-jason) [inline expression/builtin_string.go]: No, this is a bug fix, while error happens, take `from_base64("aaa")` as a example, we should return NULL.

Comment (XuHuaiyu) [inline expression/builtin_string.go]: add a comment for this may be better

Comment (zz-jason) [inline expression/builtin_string.go]: OK

**Comment by zz-jason** [comment]
@XuHuaiyu updated, PTAL again

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2126,6 +2126,7 @@ func (b *builtinFromBase64Sig) evalString(row []types.Datum) (string, bool, erro
 	str = strings.Replace(str, " ", "", -1)
 	result, err := base64.StdEncoding.DecodeString(str)
 	if err != nil {
+		// when error happens, take `from_base64("asc")` as an example, we should return NULL.
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: s/ when /When ...

File: expression/builtin_string.go [hunk 3]
```diff
@@ -2098,35 +2098,38 @@ type fromBase64FunctionClass struct {
...
-	sig := &builtinFromBase64Sig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = mysql.MaxBlobWidth
```
Comment (breeswish) [inline expression/builtin_string.go]: I think bf.tp.Flen can be inferred more accurately.

Comment (zz-jason) [inline expression/builtin_string.go]: bf.tp.Flen = arg.GetType().Flen / 4 * 3 ?

Comment (breeswish) [inline expression/builtin_string.go]: @zz-jason Yes.

**Pull request merged**